modules/programs/kakoune.nix

```diff
@@ -0,0 +1,565 @@
+          "${optionalString (wheelUpButton != null)
+            "ncurses_wheel_up_button=${wheelUpButton}"}"
+          "${optionalString (shiftFunctionKeys != null)
+            "ncurses_shift_function_key=${toString shiftFunctionKeys}"}"
+          "ncurses_builtin_key_parser=${if useBuiltinKeyParser then "true" else "false"}"
+        ];
+
+        keyMappingString = km: concatStringsSep " " [
+          "map global"
+          "${km.mode} ${km.key} '${km.effect}'"
+          "${optionalString (km.docstring != null) "-docstring '${km.docstring}'"}"
+        ];
+
+        hookString = h: concatStringsSep " " [
+          "hook" "${optionalString (h.group != null) "-group ${group}"}"
+          "${optionalString (h.once) "-once"}" "global"
+          "${h.name}" "${optionalString (h.option != null) h.option}"
+          "%{ ${h.commands} }"
+        ];
+    in  ''
+    # Generated by home-manager
+    ${optionalString (tabStop != null) "set-option global tabstop ${toString tabStop}"}
+    ${optionalString (indentWidth != null) "set-option global indentwidth ${toString indentWidth}"}
+    ${optionalString (!incrementalSearch) "set-option global incsearch false"}
+    ${optionalString (alignWithTabs) "set-option global aligntab true"}
+    ${optionalString (autoInfo != null) "set-option global autoinfo ${concatStringsSep "|" autoInfo}"}
+    ${optionalString (autoComplete != null) "set-option global autocomplete ${concatStringsSep "|" autoComplete}"}
+    ${optionalString (autoReload != null) "set-option global/ autoreload ${autoReload}"}
+    ${optionalString (wrapLines != null && wrapLines.enable) "add-highlighter global/ wrap${wrapOptions}"}
+    ${optionalString (numberLines != null && numberLines.enable)
+      "add-highlighter global/ number-lines${numberLinesOptions}"}
+    ${optionalString showMatching "add-highlighter global/ show-matching"}
+    ${optionalString (scrollOff != null)
+      "set-option global scrolloff ${toString scrollOff.lines},${toString scrollOff.columns}"}
+
+    # UI options
+    ${optionalString (ui != null) "set-option global ui_options ${uiOptions}"}
+
+    # Key mappings
+    ${concatStringsSep "\n" (map keyMappingString keyMappings)}
+
+    # Hooks
+    ${concatStringsSep "\n" (map hookString hooks)}
+  '' else "") + "\n" + cfg.extraConfig);
+
+in
+{
+  options = {
+    programs.kakoune = {
+      enable = mkEnableOption "the kakoune text editor";
+ 
+      config = mkOption {
+        type = types.nullOr configModule;
+        default = {};
+        description = "kakoune configuration options.";
+      };
+
+      extraConfig = mkOption {
+        type = types.lines;
+        default = "";
+        description = ''
+          Extra configuration lines to add to
+          <filename>~/.config/kak/kakrc</filename>.
+        '';
+      };
+    };
+  };
+
+  config = mkIf cfg.enable(mkMerge [
+    {
+      home.packages = [ pkgs.kakoune ];
+      xdg.configFile."kak/kakrc-test".source = configFile;
```

Comment:
fix the filename and we can merge since previous reviewers seem ok with the module.